### ISSUE TITLE
Add code generation support for %ROWTYPE references in Oracle

### ISSUE THREAD

**Issue by DiabolusExMachina** [jOOQ/jOOQ#5809]
@lukaseder  Hello again!

I had time time to test jOOQ 3.9 and facing some problems during code generation. We are still on Oracle 11g at the moment. 

This is my procedure

```sql
PACKAGE per_per IS

PROCEDURE read_person(
  p_person_id IN person.person_id%TYPE,
  p_rec_person  OUT person%ROWTYPE,
  p_error_proto IN OUT error_types.rec_errorprotocol
);

END per_per;
```
This procedure is part of an huge schema. So what I want is to generate only what is necessary to call this procedure. 

with 
```xml
<includes>per_per</includes>
```
I get the package and the routines in it. However in the generated class ReadPerson the parameters p_person and p_error_proto are type Object and I can't execute the procedure. 

to get the correct  rec_errorprotocol type I have to write

```xml
<includes>
 per_per
   | error_types
   | rec_errorprotocol
 </includes>
```
I'm not sure why I have to add the package and the name of the type but only if I add both the type will be generated and referenced in the generated Procedure class. (There are only types in that package nothing else...) 

Next I tried include the person type as well.
```xml
<includes>
 per_per
   | person
   | error_types
   | rec_errorprotocol
 </includes>
```
This will generate 2 classes, a Table class "Person" and a Record class "PersonRecord" but the type is not used in the generated ReadPerson class. p_person is still type Object and I still can't execute the procedure. 

am I doing something wrong?

**Comment by lukaseder** [comment]
Thanks @DiabolusExMachina for your report. I'll have to check if this behaviour is consistent / expected or if this could qualify as a bug.

In the meantime, you could always include all of a package's content by writing this instead:

```xml
<includes>per_per(\..*)?</includes>
```

This will include the package and all of its content.

**Comment by DiabolusExMachina** [comment]
Hi, 

thanks again for the fast response. 

For me there is no difference in using 
```xml
<includes>per_per</includes>
```
or 

```xml
<includes>per_per(\..*)?</includes>
```

All procedures in the package will be generated but.  p_error_proto and p_person are of type  Parameter<Object> this can be solved with using 
```xml
<includes>
 per_per(\..*)?
   | person
   | error_types(\..*)?
   | rec_errorprotocol
 </includes>
```

But still no matter what I do 

```xml
<includes>
 per_per(\..*)?
   | error_types(\..*)?
 </includes>
```
Is also not helping. The rec_errorprotocol type includes in the package specification is not generated.

**Comment by DiabolusExMachina** [comment]
Hi, 
thanks again for the fast response. There are 2 separate issues here. 

My biggest problem is that in the generated procedure Class (ReadPerson) the type of the parameter p_rec_person is  ` Parameter<Object>` 

The type itself (PersonRecord) is generated but not referenced as type of the parameter p_rec_person in the ReadPerson class, 

My second, smaller problem is that I have a package that includes only one record type in the package specification. To generate the Record I have to include the name of the package and the name of the type. This is inconvenient but manageable. If the record is generated than it is used in the ReadPerson class. 

For me there is also no difference in using  `(\..*)?`  at the end of a package. The record in the `error_types ` package is only generated if include the package and the record.

**Comment by lukaseder** [comment]
I see, thank you for detailing. Yes, I can reproduce this issue. `%ROWTYPE` types are not currently supported. That's a bug, we should reference the table's record type, or potentially generate a synthetic record type instead.

**Comment by lukaseder** [comment]
Hmm, `%ROWTYPE` support was referenced in https://github.com/jOOQ/jOOQ/issues/5627 (when support for PL/SQL types was added). This must have been an oversight.

**Comment by lukaseder** [comment]
I remember now: The problem is that `ALL_ARGUMENTS` only reports the row type structure, not the source. So, there's no link to the defining table type in `ALL_ARGUMENTS`.

However, the current query doesn't report the row type structure (to at least generate a synthetic type). This is because `%ROWTYPE` types seem not to report any `(TYPE_OWNER, TYPE_NAME, TYPE_SUBNAME)` tuple. Will investigate options

**Comment by lukaseder** [comment]
Querying the `ALL_ARGUMENTS` dictionary view yields the following findings for `%ROWTYPE`:

- There are procedure / function argument `%ROWTYPES`. They can be identified through: `OWNER.PACKAGE_NAME.OBJECT_NAME.ARGUMENT_NAME`
- There are function return `%ROWTYPE`. They can be identified through: `OWNER.PACKAGE_NAME.OBJECT_NAME` (`ARGUMENT_NAME` is null)
- There are nested `RECORD` `%ROWTYPES` (i.e. `%ROWTYPES` that are used inside of PL/SQL RECORD types). They can be identified through `OWNER.PACKAGE_NAME.OBJECT_NAME.ARGUMENT_NAME` (where `ARGUMENT_NAME` is really the `ATTR_NAME` of the containing PL/SQL RECORD type, not the procedure / function argument name (!))

**Comment by lukaseder** [comment]
This query establishes the entire child / parent relationship in `ALL_ARGUMENTS` using `SEQUENCE` and `DATA_LEVEL`:

```sql
WITH
  args AS (
    SELECT 
      a.*,
      lead(SEQUENCE, 1, 99999999) OVER (
        PARTITION BY OWNER, PACKAGE_NAME, SUBPROGRAM_ID, DATA_LEVEL
        ORDER BY SEQUENCE ASC
      ) next_sibling
    FROM SYS.ALL_ARGUMENTS a
    WHERE OWNER IN ('TEST')
    AND package_name IS NOT NULL
    --
    AND package_name = 'PLS_OBJECTS'
    AND subprogram_id = 18
  )
SELECT 
  sys_connect_by_path(sequence, '/'),
  PRIOR sequence,
  args.*
FROM args
START WITH data_level = 0
CONNECT BY data_level = PRIOR data_level + 1
AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
ORDER BY owner, object_name, package_name, subprogram_id, sequence
```

**Comment by lukaseder** [comment]
This query is work in progress:

```sql
WITH 
  args_with_siblings AS (
    SELECT 
      a.*
    FROM (
      SELECT
        lead(a.SEQUENCE, 1, 99999999) OVER (
          PARTITION BY a.OWNER, a.PACKAGE_NAME, 
                       a.SUBPROGRAM_ID, a.DATA_LEVEL
          ORDER BY a.SEQUENCE ASC
        ) next_sibling,
        a.*
      FROM SYS.ALL_ARGUMENTS a
      WHERE a.OWNER IN ('TEST')
      AND a.PACKAGE_NAME IS NOT NULL -- Optimisation here
      AND a.PACKAGE_NAME = 'PLS_OBJECTS'
      ) a
    WHERE 1 = 1
    --AND (a.TYPE_OWNER IN ('TEST') OR a.TYPE_OWNER IS NULL)
    AND a.OWNER         IN ('TEST')
    --AND a.DATA_TYPE      = 'PL/SQL RECORD'
  ),
  args_with_parent AS (
    SELECT 
      sys_connect_by_path(sequence, '/') path,
      PRIOR sequence parent_sequence,
      PRIOR argument_name parent_argument_name,
      CONNECT_BY_ROOT type_owner root_type_owner,
      CONNECT_BY_ROOT type_name root_type_name,
      CONNECT_BY_ROOT type_subname root_type_subname,
      a.*
    FROM args_with_siblings a
    START WITH DATA_TYPE = 'PL/SQL RECORD'
    CONNECT BY data_level = PRIOR data_level + 1
    AND OWNER = PRIOR owner
    AND PACKAGE_NAME = PRIOR package_name
    AND SUBPROGRAM_ID = PRIOR subprogram_id
    AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
    ORDER BY owner, object_name, package_name, subprogram_id, sequence
  )
SELECT
  CASE 
    WHEN type_owner IS NULL AND root_type_subname IS NOT NULL THEN
      owner || '.' || package_name || '.' || root_type_subname || nvl2(parent_argument_name, '.' || parent_argument_name, '')
    WHEN type_owner IS NULL AND data_type = 'PL/SQL RECORD' THEN
      owner || '.' || package_name || '.' || object_name || nvl2(argument_name, '.' || argument_name, '')
    WHEN type_owner IS NULL THEN
      owner || '.' || package_name || '.' || object_name || nvl2(parent_argument_name, '.' || parent_argument_name, '')
    ELSE
      type_owner || '.' || type_name || '.' || type_subname
  END n,
  a.*
FROM args_with_parent a
ORDER BY
  n ASC,
  sequence ASC
```

**Comment by lukaseder** [comment]
Continuing to investigate this. The proposed approach from my last comment: https://github.com/jOOQ/jOOQ/issues/5809#issuecomment-300871209 ... works well for types like these:

```sql
TYPE r IS RECORD (
  a tab%rowtype;
);

PROCEDURE proc1 (p r);
```

I.e. for nested `%rowtype` references. But it doesn't work for inline `%rowtype` references like this:

```sql
PROCEDURE proc1 (p tab%rowtype);
```

While the type *structure* can be discovered, a local variable in a generated PL/SQL block still needs to know the exact `tab%rowtype` name in order to generate something like this:

```sql
DECLARE
  -- This declaration is tricky
  v tab%rowtype;
BEGIN
  -- These assignments are easy
  v.col1 := ?;
  v.col2 := ?;
  ...
  v.colN := ?;
  -- The call is also easy
  proc1(v);
END;
```

I'm starting to feel that it won't be easy to implement this feature without actually parsing the PL/SQL code. But not all PL/SQL code is available, and parsing has its own caveats... Will need to continue looking for possible solutions.

**Comment by lukaseder** [comment]
Good news, after some further attempts, it can be shown that the `PL/SQL RECORD` type is a *structural type* with respect to the procedure call, not a *nominal type*. The following works:

```sql
DECLARE
  TYPE t IS RECORD (
    col1 NUMBER,
    col2 NUMBER,
    ...
    colN NUMBER
  );
  v t;
BEGIN
  v.col1 := ?;
  v.col2 := ?;
  ...
  v.colN := ?
  proc1(v);
END;
```

So, if we do not know the `PL/SQL RECORD` type name (i.e. when it is a top-level `tab%rowtype` reference), then we can simply reconstruct the type declaration in an anonymous block and use that instead. This shouldn't be too hard, so this feature can finally be implemented entirely for jOOQ 3.10

**Comment by lukaseder** [comment]
So, this query discovers all `PL/SQL RECORD` types including top-level or non-top-level `%rowtype`:

```sql
WITH 
  args_with_siblings AS (
    SELECT 
      a.*
    FROM (
      SELECT
        lead(a.SEQUENCE, 1, 99999999) OVER (
          PARTITION BY a.OWNER, a.PACKAGE_NAME, 
                       a.SUBPROGRAM_ID, a.DATA_LEVEL
          ORDER BY a.SEQUENCE ASC
        ) next_sibling,
        a.*
      FROM SYS.ALL_ARGUMENTS a
      WHERE a.OWNER IN ('TEST')
      AND a.PACKAGE_NAME IS NOT NULL -- Optimisation here
      AND a.PACKAGE_NAME = 'PLS_OBJECTS'
      ) a
    WHERE 1 = 1
    --AND (a.TYPE_OWNER IN ('TEST') OR a.TYPE_OWNER IS NULL)
    AND a.OWNER         IN ('TEST')
    AND a.OBJECT_NAME NOT LIKE '%ASSOC%' 
    AND a.OBJECT_NAME NOT LIKE '%TAB%'
    AND a.OBJECT_NAME NOT LIKE '%ARR%'
    --AND a.OBJECT_NAME   IN ('F_R_BOOK')
    --AND a.DATA_TYPE      = 'PL/SQL RECORD'
    ORDER BY owner, object_name, package_name, subprogram_id, sequence
  ),
  top_level_rowtypes AS (
    SELECT 
      'TOP_LEVEL_ROWTYPE' type,
      level lvl,
      subprogram_id || sys_connect_by_path(lpad(to_char(sequence), 3, '000'), '/') path,
      owner || '.' || package_name || '.' || object_name || nvl2(PRIOR argument_name, '.' || PRIOR argument_name, '') qualifier,
      a.*
    FROM args_with_siblings a
    START WITH data_type = 'PL/SQL RECORD' 
      AND type_owner IS NULL AND type_name IS NULL AND type_subname IS NULL
      AND data_level = 0
    CONNECT BY data_level = PRIOR data_level + 1
      AND owner = PRIOR owner AND package_name = PRIOR package_name AND subprogram_id = PRIOR subprogram_id
      AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
  ),
  nested_rowtypes AS (
    SELECT
      'NESTED_ROWTYPE' type,
      level lvl,
      subprogram_id || sys_connect_by_path(lpad(to_char(sequence), 3, '000'), '/') path,
      (
        SELECT type_owner || '.' || type_name || '.' || type_subname
        FROM args_with_siblings p
        WHERE a.data_level = p.data_level + 2
          AND a.owner = p.owner AND a.package_name = p.package_name AND a.subprogram_id = p.subprogram_id
          AND a.sequence BETWEEN p.sequence AND p.next_sibling
      ) || '.' || PRIOR argument_name qualifier,
      a.*
    FROM args_with_siblings a
    START WITH data_type = 'PL/SQL RECORD'
      AND type_owner IS NULL AND type_name IS NULL AND type_subname IS NULL
      AND data_level > 0
    CONNECT BY data_level = PRIOR data_level + 1
      AND owner = PRIOR owner AND package_name = PRIOR package_name AND subprogram_id = PRIOR subprogram_id
      AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
  ),
  non_rowtypes AS (
    SELECT
      'NON_ROWTYPE' type,
      level lvl,
      subprogram_id || sys_connect_by_path(lpad(to_char(sequence), 3, '000'), '/') path,
      PRIOR type_owner || '.' || PRIOR type_name || '.' || PRIOR type_subname qualifier,
      a.*
    FROM args_with_siblings a
    START WITH data_type = 'PL/SQL RECORD'
      AND type_owner IS NOT NULL AND type_name IS NOT NULL AND type_subname IS NOT NULL
    CONNECT BY data_level = PRIOR data_level + 1
      AND owner = PRIOR owner AND package_name = PRIOR package_name AND subprogram_id = PRIOR subprogram_id
      AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
  )
SELECT *
FROM top_level_rowtypes
WHERE lvl = 2
UNION ALL
SELECT *
FROM nested_rowtypes
WHERE lvl = 2
UNION ALL
SELECT *
FROM non_rowtypes
WHERE lvl = 2
ORDER BY 3
```

**Comment by lukaseder** [comment]
Faster query with factored out recursion:

```sql
WITH 
  args_with_siblings AS (
    SELECT 
      a.*
    FROM (
      SELECT
        lead(a.SEQUENCE, 1, 99999999) OVER (
          PARTITION BY a.OWNER, a.PACKAGE_NAME, 
                       a.SUBPROGRAM_ID, a.DATA_LEVEL
          ORDER BY a.SEQUENCE ASC
        ) next_sibling,
        a.*
      FROM SYS.ALL_ARGUMENTS a
      WHERE a.OWNER IN ('TEST')
      AND a.PACKAGE_NAME IS NOT NULL -- Optimisation here
      AND a.PACKAGE_NAME = 'PLS_OBJECTS'
      ) a
    WHERE 1 = 1
    --AND (a.TYPE_OWNER IN ('TEST') OR a.TYPE_OWNER IS NULL)
    AND a.OWNER         IN ('TEST')
    AND a.OBJECT_NAME NOT LIKE '%ASSOC%' 
    AND a.OBJECT_NAME NOT LIKE '%TAB%'
    AND a.OBJECT_NAME NOT LIKE '%ARR%'
    --AND a.OBJECT_NAME   IN ('F_R_BOOK')
    --AND a.DATA_TYPE      = 'PL/SQL RECORD'
    ORDER BY owner, object_name, package_name, subprogram_id, sequence
  ),
  record_types AS (
    SELECT 
      CASE 
        WHEN CONNECT_BY_ROOT type_owner IS NULL AND CONNECT_BY_ROOT type_name IS NULL AND CONNECT_BY_ROOT type_subname IS NULL THEN 
          CASE 
            WHEN CONNECT_BY_ROOT data_level = 0 THEN 'TOP_LEVEL_ROWTYPE' 
                                                ELSE 'NESTED_ROWTYPE'   
          END
        ELSE 'NON_ROWTYPE'
      END AS type,
      level lvl,
      subprogram_id || sys_connect_by_path(lpad(to_char(sequence), 3, '000'), '/') path,
      CASE 
        WHEN CONNECT_BY_ROOT type_owner IS NULL AND CONNECT_BY_ROOT type_name IS NULL AND CONNECT_BY_ROOT type_subname IS NULL THEN 
          CASE 
            WHEN CONNECT_BY_ROOT data_level = 0 THEN owner || '.' || package_name || '.' || object_name || nvl2(PRIOR argument_name, '.' || PRIOR argument_name, '')
                                                ELSE (
              SELECT type_owner || '.' || type_name || '.' || type_subname
              FROM args_with_siblings p
              WHERE a.data_level = p.data_level + 2
                AND a.owner = p.owner AND a.package_name = p.package_name AND a.subprogram_id = p.subprogram_id
                AND a.sequence BETWEEN p.sequence AND p.next_sibling
            ) || '.' || PRIOR argument_name
          END
        ELSE PRIOR type_owner || '.' || PRIOR type_name || '.' || PRIOR type_subname
      END AS qualifier,
      a.*
    FROM args_with_siblings a
    START WITH data_type = 'PL/SQL RECORD' 
    CONNECT BY data_level = PRIOR data_level + 1
      AND owner = PRIOR owner AND package_name = PRIOR package_name AND subprogram_id = PRIOR subprogram_id
      AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
  )
SELECT *
FROM record_types
WHERE lvl = 2
ORDER BY path
```

**Comment by lukaseder** [comment]
One more step, the final query removes distinct type references in case non `%rowtype` types are referenced multiple times from different stored procedures:

```sql
WITH 
  args_with_siblings AS (
    SELECT 
      a.*
    FROM (
      SELECT
        lead(a.SEQUENCE, 1, 99999999) OVER (
          PARTITION BY a.OWNER, a.PACKAGE_NAME, 
                       a.SUBPROGRAM_ID, a.DATA_LEVEL
          ORDER BY a.SEQUENCE ASC
        ) next_sibling,
        a.*
      FROM SYS.ALL_ARGUMENTS a
      WHERE a.OWNER IN ('TEST')
      AND a.PACKAGE_NAME IS NOT NULL -- Optimisation here
      AND a.PACKAGE_NAME = 'PLS_OBJECTS'
      ) a
    WHERE 1 = 1
    AND a.OWNER         IN ('TEST')
    ORDER BY owner, object_name, package_name, subprogram_id, sequence
  ),
  record_types AS (
    SELECT 
      CASE 
        WHEN CONNECT_BY_ROOT type_owner IS NULL AND CONNECT_BY_ROOT type_name IS NULL AND CONNECT_BY_ROOT type_subname IS NULL THEN 
          CASE 
            WHEN CONNECT_BY_ROOT data_level = 0 THEN 'TOP_LEVEL_ROWTYPE' 
                                                ELSE 'NESTED_ROWTYPE'   
          END
        ELSE 'NON_ROWTYPE'
      END AS type,
      level lvl,
      subprogram_id || sys_connect_by_path(lpad(to_char(sequence), 3, '000'), '/') path,
      CASE 
        WHEN CONNECT_BY_ROOT type_owner IS NULL AND CONNECT_BY_ROOT type_name IS NULL AND CONNECT_BY_ROOT type_subname IS NULL THEN 
          CASE 
            WHEN CONNECT_BY_ROOT data_level = 0 THEN owner || '.' || package_name || '.' || object_name || nvl2(PRIOR argument_name, '.' || PRIOR argument_name, '')
                                                ELSE (
              SELECT type_owner || '.' || type_name || '.' || type_subname
              FROM args_with_siblings p
              WHERE a.data_level = p.data_level + 2
                AND a.owner = p.owner AND a.package_name = p.package_name AND a.subprogram_id = p.subprogram_id
                AND a.sequence BETWEEN p.sequence AND p.next_sibling
            ) || '.' || PRIOR argument_name
          END
        ELSE PRIOR type_owner || '.' || PRIOR type_name || '.' || PRIOR type_subname
      END AS qualifier,
      a.*
    FROM args_with_siblings a
    START WITH data_type = 'PL/SQL RECORD' 
    CONNECT BY data_level = PRIOR data_level + 1
      AND owner = PRIOR owner AND package_name = PRIOR package_name AND subprogram_id = PRIOR subprogram_id
      AND sequence BETWEEN PRIOR sequence AND PRIOR next_sibling
  )
SELECT DISTINCT
  type,
  qualifier,
  argument_name,
  position,
  type_owner,
  nvl2(type_subname, type_name, null),
  coalesce(type_subname, type_name, data_type),
  data_length,
  data_precision,
  data_scale
FROM record_types
WHERE lvl = 2
ORDER BY
  qualifier,
  position
```

**Comment by lukaseder** [comment]
The query is far too slow on the `SYS` schema. It ran in 50 seconds on Oracle 12cR1. Let's see if the query can be simplified by using a JOIN. After all, only 2 levels of recursion are ever needed.

**Comment by lukaseder** [comment]
Here's again the query with a JOIN:

```sql
WITH
  args_with_siblings AS (
    SELECT
      lead(a.SEQUENCE, 1, 99999999) OVER (
        PARTITION BY a.OWNER, a.PACKAGE_NAME, 
                     a.SUBPROGRAM_ID, a.DATA_LEVEL
        ORDER BY a.SEQUENCE ASC
      ) next_sibling,
      a.*
    FROM SYS.ALL_ARGUMENTS a
    WHERE a.OWNER IN ('TEST', 'SYS')
    AND a.PACKAGE_NAME IS NOT NULL -- Optimisation here
    AND a.PACKAGE_NAME = 'PLS_OBJECTS'
  ),
  record_types AS (
    SELECT 
      CASE 
        WHEN p.type_owner IS NULL AND p.type_name IS NULL AND p.type_subname IS NULL THEN 
          CASE 
            WHEN p.data_level = 0 THEN 'TOP_LEVEL_ROWTYPE' 
                                  ELSE 'NESTED_ROWTYPE'   
          END
        ELSE 'NON_ROWTYPE'
      END AS type,
      CASE 
        WHEN p.type_owner IS NULL AND p.type_name IS NULL AND p.type_subname IS NULL THEN 
          CASE 
            WHEN p.data_level = 0 THEN a.owner || '.' || a.package_name || '.' || a.object_name || nvl2(p.argument_name, '.' || p.argument_name, '')
                                  ELSE (
              SELECT pp.type_owner || '.' || pp.type_name || '.' || pp.type_subname
              FROM args_with_siblings pp
              WHERE a.data_level = pp.data_level + 2
                AND a.owner = pp.owner AND a.package_name = pp.package_name AND a.subprogram_id = pp.subprogram_id
                AND a.sequence BETWEEN pp.sequence AND pp.next_sibling
            ) || '.' || p.argument_name
          END
        ELSE p.type_owner || '.' || p.type_name || '.' || p.type_subname
      END AS qualifier,
      a.*
    FROM args_with_siblings a
    JOIN args_with_siblings p
      ON p.data_type = 'PL/SQL RECORD'
      AND a.data_level = p.data_level + 1
      AND a.owner = p.owner
      AND a.package_name = p.package_name
      AND a.subprogram_id = p.subprogram_id
      AND a.sequence BETWEEN p.sequence and p.next_sibling
  )
SELECT DISTINCT
  type,
  qualifier,
  argument_name,
  position,
  type_owner,
  nvl2(type_subname, type_name, null),
  coalesce(type_subname, type_name, data_type),
  data_length,
  data_precision,
  data_scale
FROM record_types
ORDER BY
  qualifier,
  position
```

**Comment by lukaseder** [comment]
Finally, the jOOQ generated query:

```sql
with 
  "args_with_siblings" as (
    select 
      "a"."OWNER", 
      "a"."OBJECT_NAME", 
      "a"."PACKAGE_NAME", 
      "a"."OBJECT_ID", 
      "a"."OVERLOAD", 
      "a"."SUBPROGRAM_ID", 
      "a"."ARGUMENT_NAME", 
      "a"."POSITION", 
      "a"."SEQUENCE", 
      "a"."DATA_LEVEL", 
      "a"."DATA_TYPE", 
      "a"."DEFAULTED", 
      "a"."DEFAULT_VALUE", 
      "a"."DEFAULT_LENGTH", 
      "a"."IN_OUT", 
      "a"."DATA_LENGTH", 
      "a"."DATA_PRECISION", 
      "a"."DATA_SCALE", 
      "a"."RADIX", 
      "a"."CHARACTER_SET_NAME", 
      "a"."TYPE_OWNER", 
      "a"."TYPE_NAME", 
      "a"."TYPE_SUBNAME", 
      "a"."TYPE_LINK", 
      "a"."PLS_TYPE", 
      "a"."CHAR_LENGTH", 
      "a"."CHAR_USED", 
      "a"."ORIGIN_CON_ID", 
      lead(
        "a"."SEQUENCE", 
        1, 
        99999999
      ) over (partition by 
        "a"."OWNER", 
        "a"."PACKAGE_NAME", 
        "a"."SUBPROGRAM_ID", 
        "a"."DATA_LEVEL"
       order by "a"."SEQUENCE") "next_sibling"
    from "SYS"."ALL_ARGUMENTS" "a"
    where (
      "a"."OWNER" in (
        'TEST'
      )
      and "a"."PACKAGE_NAME" is not null
    )
  ), 
  "record_types" as (
    select 
      case when (
            "p"."TYPE_OWNER" is null
            and "p"."TYPE_NAME" is null
            and "p"."TYPE_SUBNAME" is null
          ) then case when "p"."DATA_LEVEL" = 0 then 'TOP_LEVEL_ROWTYPE'
                      else 'NESTED_ROWTYPE'
                 end
           else 'NON_ROWTYPE'
      end "type", 
      case when (
            "p"."TYPE_OWNER" is null
            and "p"."TYPE_NAME" is null
            and "p"."TYPE_SUBNAME" is null
          ) then case when "p"."DATA_LEVEL" = 0 then ((((("a"."OWNER" || '.') || "a"."PACKAGE_NAME") || '.') || "a"."OBJECT_NAME") || nvl2("p"."ARGUMENT_NAME", ('_' || "p"."ARGUMENT_NAME"), ''))
                      else ((
                       select (((("q"."TYPE_OWNER" || '.') || "q"."TYPE_NAME") || '.') || "q"."TYPE_SUBNAME")
                       from "args_with_siblings" "q"
                       where (
                         "a"."DATA_LEVEL" = ("q"."DATA_LEVEL" + 2)
                         and "a"."OWNER" = "q"."OWNER"
                         and "a"."PACKAGE_NAME" = "q"."PACKAGE_NAME"
                         and "a"."SUBPROGRAM_ID" = "q"."SUBPROGRAM_ID"
                         and "a"."SEQUENCE" between "q"."SEQUENCE" and "q"."next_sibling"
                       )
                     ) || nvl2("p"."ARGUMENT_NAME", ('_' || "p"."ARGUMENT_NAME"), ''))
                 end
           else (((("p"."TYPE_OWNER" || '.') || "p"."TYPE_NAME") || '.') || "p"."TYPE_SUBNAME")
      end "qualifier", 
      "a"."OWNER", 
      "a"."OBJECT_NAME", 
      "a"."PACKAGE_NAME", 
      "a"."OBJECT_ID", 
      "a"."OVERLOAD", 
      "a"."SUBPROGRAM_ID", 
      "a"."ARGUMENT_NAME", 
      "a"."POSITION", 
      "a"."SEQUENCE", 
      "a"."DATA_LEVEL", 
      "a"."DATA_TYPE", 
      "a"."DEFAULTED", 
      "a"."DEFAULT_VALUE", 
      "a"."DEFAULT_LENGTH", 
      "a"."IN_OUT", 
      "a"."DATA_LENGTH", 
      "a"."DATA_PRECISION", 
      "a"."DATA_SCALE", 
      "a"."RADIX", 
      "a"."CHARACTER_SET_NAME", 
      "a"."TYPE_OWNER", 
      "a"."TYPE_NAME", 
      "a"."TYPE_SUBNAME", 
      "a"."TYPE_LINK", 
      "a"."PLS_TYPE", 
      "a"."CHAR_LENGTH", 
      "a"."CHAR_USED", 
      "a"."ORIGIN_CON_ID"
    from "args_with_siblings" "a"
      join "args_with_siblings" "p"
      on (
        "p"."DATA_TYPE" = 'PL/SQL RECORD'
        and "a"."DATA_LEVEL" = ("p"."DATA_LEVEL" + 1)
        and "a"."OWNER" = "p"."OWNER"
        and "a"."PACKAGE_NAME" = "p"."PACKAGE_NAME"
        and "a"."SUBPROGRAM_ID" = "p"."SUBPROGRAM_ID"
        and "a"."SEQUENCE" between "p"."SEQUENCE" and "p"."next_sibling"
      )
  )
select distinct 
  "qualifier", 
  "a"."ARGUMENT_NAME" "ATTR_NAME", 
  "a"."POSITION" "ATTR_NO", 
  case when (
        "a"."DATA_TYPE" = 'PL/SQL RECORD'
        and "a"."TYPE_OWNER" is null
      ) then "a"."OWNER"
       else "a"."TYPE_OWNER"
  end "ATTR_TYPE_OWNER", 
  case when (
        "a"."DATA_TYPE" = 'PL/SQL RECORD'
        and "a"."TYPE_OWNER" is null
      ) then "a"."PACKAGE_NAME"
       else nvl2("a"."TYPE_SUBNAME", "a"."TYPE_NAME", null)
  end "package_name", 
  case when (
        "a"."DATA_TYPE" = 'PL/SQL RECORD'
        and "a"."TYPE_OWNER" is null
      ) then ((substr(
        "qualifier", 
        instr("qualifier", '.', 1, 2) + 1
      ) || '_') || "a"."ARGUMENT_NAME")
       else coalesce(
        "a"."TYPE_SUBNAME", 
        "a"."TYPE_NAME", 
        "a"."DATA_TYPE"
      )
  end "ATTR_TYPE_NAME", 
  "a"."DATA_LENGTH" "LENGTH", 
  "a"."DATA_PRECISION" "PRECISION", 
  "a"."DATA_SCALE" "SCALE"
from "record_types" "a"
order by 
  "qualifier", 
  "a"."POSITION"
```

**Comment by lukaseder** [comment]
Great news. Code generation support for `%ROWTYPE` references is now implemented in jOOQ 3.10.

Some tasks still left:

- Runtime support for calling functions / procedures: #6530
- Support for overloaded methods contain top-level `%ROWTYPE` references

**Comment by lukaseder** [comment]
It's been a few years, so maybe, we can now support `%ROWTYPE` references in our anonymous PL/SQL blocks for more recent Oracle versions (19c or newer). I'll investigate this:

- https://github.com/jOOQ/jOOQ/issues/17369
